File: Maths&Geometry/Leetcode/Hard/ErectTheFence/ErectTheFence.java
// https://leetcode.com/problems/erect-the-fence/
import java.util.*;
public class ErectTheFence {
    class Pair {
        int first, second;
        Pair(int first, int second) {
            this.first = first;
            this.second = second;
        }

        @Override
        public boolean equals(Object b) {
            if(!(b instanceof Pair)) return false;
            Pair b1 = (Pair)b;
            return this.first == b1.first && this.second == b1.second;
        }

        @Override
        public int hashCode() {
            return Objects.hash(first, second);
        }
    }
    int equation(Pair p1, Pair p2, Pair p3) {
        // m23 - m21 > 0 (for convex in lower half)
        // (y3-y2)/(x3-x2) - (y2-y1)/(x2-x1)
        // (y3-y2)(x2-x1) - (y2-y1)(x3-x2)
        int x1 = p1.first, x2 = p2.first, x3 = p3.first;
        int y1 = p1.second, y2 = p2.second, y3 = p3.second;

        return (y3-y2)*(x2-x1) - (y2-y1)*(x3-x2);
    }
    public int[][] outerTrees(int[][] trees) {

        List<Pair> lower = new ArrayList<>(), upper = new ArrayList<>();

        // looking from left bottom to right
        Arrays.sort(trees, (a, b) -> {
            if(a[0] == b[0]) return a[1] - b[1];
            return a[0] - b[0];
        });

        for(int[] point: trees) {
            int l = lower.size(), u = upper.size();
            // positive for lower negative for above for convexity
            while(l >= 2 && equation(lower.get(l-2), lower.get(l-1), new Pair(point[0], point[1])) < 0) {
                l--;
                lower.remove(l);
            }
            while(u >= 2 && equation(upper.get(u-2), upper.get(u-1), new Pair(point[0], point[1])) > 0) {
                u--;
                upper.remove(u);
            }

            lower.add(new Pair(point[0], point[1]));
            upper.add(new Pair(point[0], point[1]));
        }

        Set<Pair> st = new HashSet<>();

        for(Pair point: lower) {
            st.add(point);
        }
        for(Pair point: upper) {
            st.add(point);
        }

        int[][] ans = new int[st.size()][2];
        int ind = 0;
        for(Pair point: st) {
            ans[ind][0] = point.first;
            ans[ind][1] = point.second;
            ind++;
        }

        return ans;
    }
}

File: Maths&Geometry/Leetcode/Hard/ErectTheFence/ErectTheFence.cpp
// https://leetcode.com/problems/erect-the-fence/
#include <bits/stdc++.h>
using namespace std;

int equation(pair<int, int> &p1, pair<int, int> &p2, pair<int, int> p3)
{
    // m23 - m21 > 0 (for convex in lower half)
    // (y3-y2)/(x3-x2) - (y2-y1)/(x2-x1)
    // (y3-y2)(x2-x1) - (y2-y1)(x3-x2)
    int x1 = p1.first, y1 = p1.second, x2 = p2.first, y2 = p2.second, x3 = p3.first, y3 = p3.second;

    return (y3 - y2) * (x2 - x1) - (y2 - y1) * (x3 - x2);
}
vector<vector<int>> outerTrees(vector<vector<int>> &trees)
{
    // looking from left bottom to right
    sort(trees.begin(), trees.end());

    deque<pair<int, int>> lower, upper;

    for (auto &point : trees)
    {
        // positive for lower negative for above for convexity
        while (lower.size() >= 2 && equation(lower[lower.size() - 2], lower[lower.size() - 1], {point[0], point[1]}) < 0)
        {
            lower.pop_back();
        }

        while (upper.size() >= 2 && equation(upper[upper.size() - 2], upper[upper.size() - 1], {point[0], point[1]}) > 0)
        {
            upper.pop_back();
        }

        lower.push_back({point[0], point[1]});
        upper.push_back({point[0], point[1]});
    }

    set<pair<int, int>> st;

    for (auto &point : lower)
    {
        st.insert(point);
    }
    for (auto &point : upper)
    {
        st.insert(point);
    }

    vector<vector<int>> ans;
    for (auto &point : st)
    {
        ans.push_back({point.first, point.second});
    }

    return ans;
}